File: TiAires/app/src/main/java/net/teamtruta/tiaires/GeoCacheDetailActivity.java
package net.teamtruta.tiaires;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.SoundPool;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.RelativeLayout;

import androidx.appcompat.widget.Toolbar;
import androidx.core.content.FileProvider;

import com.microsoft.appcenter.analytics.Analytics;

import net.teamtruta.tiaires.data.GeoCache;
import net.teamtruta.tiaires.data.GeoCacheInTourWithDetails;
import net.teamtruta.tiaires.data.GeoCacheInTour;
import net.teamtruta.tiaires.viewModels.TourViewModel;
import net.teamtruta.tiaires.viewModels.TourViewModelFactory;

import org.honorato.multistatetogglebutton.MultiStateToggleButton;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

public class GeoCacheDetailActivity extends AppCompatActivity implements PopupMenu.OnMenuItemClickListener {

    GeoCacheInTour currentGeoCacheInTour;

    SoundPool soundPool;
    int soundID;

    long _tourID;

    static final int REQUEST_IMAGE_CAPTURE = 1;
    String currentPhotoPath;
    String TAG = GeoCacheDetailActivity.class.getSimpleName();

    TourViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_geo_cache_detail);

        Toolbar toolbar = findViewById(R.id.toolbar_geo_cache_detail);
        setSupportActionBar(toolbar);


        Intent intent = getIntent();
        long geoCacheID = intent.getLongExtra(App.GEOCACHE_ID_EXTRA,  -1L);
        _tourID = intent.getLongExtra(App.TOUR_ID_EXTRA, -1L);

        /*if(geoCacheID == -1L){
            // TODO: Something went wrong
        } else {

        }*/

        // Setup Action Bar
        ActionBar ab = getSupportActionBar();
        assert ab != null;
        ab.setDisplayHomeAsUpEnabled(true);

        // Get ViewModel
        viewModel = new TourViewModelFactory(((App) getApplication()).getRepository())
                .create(TourViewModel.class);

        // Observe geocache that was clicked on
        viewModel.getGeoCacheInTourFromID(geoCacheID).observe(this,
                this::setupCacheSpecificDate
                );


        //  Setup ping sound
        setupAudio();

    }

    void setupCacheSpecificDate(GeoCacheInTourWithDetails geoCacheInTour){

        currentGeoCacheInTour = geoCacheInTour.getGeoCacheInTour();
        GeoCache currentGeoCache = geoCacheInTour.getGeoCache().getGeoCache();

        // Set GeoCache Title
        ActionBar ab = getSupportActionBar();
        ab.setTitle(currentGeoCache.getName());


        // Set Not Found / Found / DNF toggle and appropriate onClickListener
        MultiStateToggleButton geoCacheVisitButton = this.findViewById(R.id.geo_cache_visit_button);
        boolean[] buttonStates = new boolean[] {true, false, false};

        if(this.currentGeoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Found)
            buttonStates = new boolean[] {false, true, false};
        else if(this.currentGeoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.DNF)
            buttonStates = new boolean[] {false, false, true};

        geoCacheVisitButton.setStates(buttonStates);

        geoCacheVisitButton.setOnValueChangedListener(position -> {
            if(position == 0){
                this.currentGeoCacheInTour.setCurrentVisitOutcome(VisitOutcomeEnum.NotAttempted);
            }
            else if(position == 1) geoCacheFound();
            else if(position == 2) geoCacheNotFound();
        });

        // Set Checkboxes
        // 1. Needs Maintenace Checkbox
        CheckBox needsMaintenanceCheckBox = findViewById(R.id.needsMaintenanceCheckBox);
        if(this.currentGeoCacheInTour.getNeedsMaintenance()) needsMaintenanceCheckBox.setChecked(true);

        // 2. FoundTrackable Checkbox and EditText
        CheckBox foundTrackableCheckBox = findViewById(R.id.foundTrackableCheckBox);
        if(currentGeoCacheInTour.getFoundTrackable() != null){
            foundTrackableCheckBox.setChecked(true);
            EditText foundTrackableEditText = findViewById(R.id.foundTrackableEditText);
            foundTrackableEditText.setText(currentGeoCacheInTour.getFoundTrackable());
        }

        // 3. DroppedTrackable Checkbox and EditText
        CheckBox droppedTrackableCheckBox = findViewById(R.id.droppedTrackableCheckBox);
        if(currentGeoCacheInTour.getDroppedTrackable() != null){
            droppedTrackableCheckBox.setChecked(true);
            EditText droppedTrackableEditText = findViewById(R.id.droppedTrackableEditText);
            droppedTrackableEditText.setText(currentGeoCacheInTour.getDroppedTrackable());
        }

        // 4. Favourite Point Checkbox
        CheckBox favouritePointCheckBox = findViewById(R.id.favouritePointCheckBox);
        if(this.currentGeoCacheInTour.getFavouritePoint()) favouritePointCheckBox.setChecked(true);

        // Set my notes
        EditText notesSection = findViewById(R.id.notes);
        String myNotes = this.currentGeoCacheInTour.getNotes();
        if(!myNotes.equals("")) notesSection.setText(myNotes);

        // Setup photo
        CheckBox photoCheckBox = findViewById(R.id.photo_checkbox);
        photoCheckBox.setChecked(currentGeoCacheInTour.getPathToImage() != null);
    }

    @Override
    public boolean onSupportNavigateUp() {
        //onBackPressed();

        Intent intent = new Intent(this, TourActivity.class);
        intent.putExtra(App.TOUR_ID_EXTRA, _tourID);
        //startActivity(intent);
        setResult(RESULT_OK, intent);
        finish();

        return true;
    }

    @Override
    public void onBackPressed() {
        saveChanges();
        onSupportNavigateUp();
    }

    public void geoCacheFound(){

        // Geocache was found

        if(currentGeoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.Found){
            // If geocache has already been found and we are clicking on Found again
            // We want to reverse this -- set geocache as not Attempted
            currentGeoCacheInTour.setCurrentVisitOutcome(VisitOutcomeEnum.NotAttempted);
            currentGeoCacheInTour.setCurrentVisitDatetime(null);

            MultiStateToggleButton geoCacheVisitButton = this.findViewById(R.id.geo_cache_visit_button);
            geoCacheVisitButton.setStates(new boolean[] {true, false, false});

        } else {
            // We want to set this geocache as found
            currentGeoCacheInTour.setCurrentVisitOutcome(VisitOutcomeEnum.Found);
            currentGeoCacheInTour.setCurrentVisitDatetime(Instant.now());
            playPing();
        }


    }

    public void geoCacheNotFound(){
        // DNF

        if(currentGeoCacheInTour.getCurrentVisitOutcome() == VisitOutcomeEnum.DNF){
            // If geocache is already a DNF and we are clicking on DNF again
            // We want to reverse this -- set geocache as not Attempted
            currentGeoCacheInTour.setCurrentVisitOutcome(VisitOutcomeEnum.NotAttempted);
            currentGeoCacheInTour.setCurrentVisitDatetime(null);

            MultiStateToggleButton geoCacheVisitButton = this.findViewById(R.id.geo_cache_visit_button);
            geoCacheVisitButton.setStates(new boolean[] {true, false, false});

        } else {
            // We want to set this geoCache as DNF
            currentGeoCacheInTour.setCurrentVisitOutcome(VisitOutcomeEnum.DNF);
            currentGeoCacheInTour.setCurrentVisitDatetime(Instant.now());

            playPing();
        }

    }

    /**
     * Handle Saving in the geoCache in tour activity
     */
    public void saveChanges()
    {
        Analytics.trackEvent("GeoCacheDetailActivity.saveChanges");

        // Get notes
        EditText notesView = findViewById(R.id.notes);
        String myNotes = notesView.getText().toString();
        currentGeoCacheInTour.setNotes(myNotes);

        // Get trackable information
        EditText foundTrackableEditText = findViewById(R.id.foundTrackableEditText);
        String foundTrackableString = foundTrackableEditText.getText().toString().trim();
        currentGeoCacheInTour.setFoundTrackable(foundTrackableString.isEmpty() ? null : foundTrackableString);

        EditText droppedTrackableEditText = findViewById(R.id.droppedTrackableEditText);
        String droppedTrackableString = droppedTrackableEditText.getText().toString().trim();
        currentGeoCacheInTour.setDroppedTrackable(droppedTrackableString.isEmpty() ? null : droppedTrackableString);

        // save changes
        viewModel.updateGeoCacheInTour(currentGeoCacheInTour);
    }

    public void onNeedsMaintenanceCheckboxClicked(View view) {

        CheckBox needsMaintenanceCheckBox = findViewById(R.id.needsMaintenanceCheckBox);
        currentGeoCacheInTour.setNeedsMaintenance(needsMaintenanceCheckBox.isChecked());

    }

    public void onFoundTrackableCheckboxClicked(View view) {

        CheckBox foundTrackableCheckBox = findViewById(R.id.foundTrackableCheckBox);
        EditText foundTrackableEditText = findViewById(R.id.foundTrackableEditText);

        if(foundTrackableCheckBox.isChecked()){

            // Focus attention on text box to fill in value
            foundTrackableEditText.requestFocus();

            //currentGeoCache.setFoundTrackable(true);
        }

        else{
            currentGeoCacheInTour.setFoundTrackable(null);

            // Delete inputted trackable code in editText area
            foundTrackableEditText.setText("");
        }

    }

    public void onDroppedTrackableCheckboxClicked(View view) {

        CheckBox droppedTrackableCheckBox = findViewById(R.id.droppedTrackableCheckBox);
        EditText droppedTrackableEditText = findViewById(R.id.droppedTrackableEditText);

        if(droppedTrackableCheckBox.isChecked()){
            //currentGeoCache.setDroppedTrackable(true);

            // Focus attention on text box to fill in value
            droppedTrackableEditText.requestFocus();
        }

        else{
            currentGeoCacheInTour.setDroppedTrackable(null);
            droppedTrackableEditText.setText("");
        }


    }

    public void onFavouritePointCheckboxClicked(View view) {

        CheckBox favouritePointCheckBox = findViewById(R.id.favouritePointCheckBox);
        currentGeoCacheInTour.setFavouritePoint(favouritePointCheckBox.isChecked());

    }

    @Override
    protected void onPause(){
        saveChanges();
        super.onPause();
    }

    void playPing(){
        // AudioManager audio settings for adjusting the volume
        AudioManager audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        float volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
        soundPool.play(soundID, volume, volume, 1, 0, 1f);
    }

    void setupAudio(){
        AudioAttributes attributes = new AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_GAME)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build();

        soundPool = new SoundPool.Builder()
                .setAudioAttributes(attributes)
                .build();

        soundID = soundPool.load(this, R.raw.ping, 1);


    }

    public void takePhoto(View view){
        takePhoto();
    }

    public void takePhoto(){
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
            // Create the File where the photo should go
            File photoFile = null;
            try {
                photoFile = createImageFile();
            } catch (IOException ex) {
                // TODO: Error occurred while creating the File
                ex.printStackTrace();
            }
            // Continue only if the File was successfully created
            if (photoFile != null) {
                Uri photoURI = FileProvider.getUriForFile(this,
                        "net.teamtruta.tiaires.fileprovider",
                        photoFile);
                Log.d(TAG, String.valueOf(photoURI));
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == RESULT_OK) {
            galleryAddPic();
        }
    }

    private File createImageFile() throws IOException {
        // Create an image file name
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
                .format(new Date());
        String imageFileName = "GEOCACHEID_" + currentGeoCacheInTour.getId() + "_" + timeStamp + "_";
        File storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File image = File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",         /* suffix */
                storageDir      /* directory */
        );

        // Save a file: path for use with ACTION_VIEW intents
        currentPhotoPath = image.getAbsolutePath();
        currentGeoCacheInTour.setPathToImage(currentPhotoPath);

        return image;
    }

    private void galleryAddPic() {
        //invoke the system's media scanner to add your photo to the Media Provider's database,
        // making it available in the Android Gallery application and to other apps.
        Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        File f = new File(currentPhotoPath);
        Uri contentUri = Uri.fromFile(f);
        mediaScanIntent.setData(contentUri);
        this.sendBroadcast(mediaScanIntent);

    }

    public void showPhotoPopup(View view){

        // If the current geoCache has no photo then go straight into the photo taking
        if(currentGeoCacheInTour.getPathToImage() == null){
            takePhoto();
            // Set the checkbox to checked
            CheckBox photoCheckBox = findViewById(R.id.photo_checkbox);
            photoCheckBox.setChecked(true);
            return;
        }

        // Set the checkbox to checked
        CheckBox photoCheckBox = findViewById(R.id.photo_checkbox);
        photoCheckBox.setChecked(true);

        PopupMenu popup = new PopupMenu(this, view);
        // This activity implements OnMenuItemClickListener
        popup.setOnMenuItemClickListener(this);
        popup.inflate(R.menu.photo_actions);
        popup.show();
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.take_new_photo){
            takePhoto();
            return true;
        }
        if (id == R.id.show_geo_cache_photo){
            showGeoCachePhoto();
            return true;
        }
        if(id == R.id.delete_geo_cache_photo){
            deleteGeoCachePhoto();
            return true;
        }
        return false;
    }
    private void deleteGeoCachePhoto() {
        File file = new File(Objects.requireNonNull(currentGeoCacheInTour.getPathToImage()));

        if (file.exists()) {
            if (file.delete()) {
                System.out.println("File Deleted :" );
            } else {
                System.out.println("File not Deleted :");
            }
        }
        currentGeoCacheInTour.setPathToImage(null);
        CheckBox photoCheckbox = findViewById(R.id.photo_checkbox);
        photoCheckbox.setChecked(false);
    }

/*  // Throws an exception because it exposes App files to the system
    void showCachePhoto(){
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_VIEW);
        intent.setDataAndType(Uri.parse("file://" + currentGeoCache.getPathToImage()), "image/*");
        startActivity(intent);
    }
*/

    public void showGeoCachePhoto() {
        Dialog builder = new Dialog(this);
        builder.requestWindowFeature(Window.FEATURE_NO_TITLE);
        builder.getWindow().setBackgroundDrawable(
                new ColorDrawable(android.graphics.Color.TRANSPARENT));
        builder.setOnDismissListener(dialogInterface -> {
            //nothing;
        });

        ImageView imageView = new ImageView(this);
        imageView.setImageURI(Uri.parse(currentGeoCacheInTour.getPathToImage()));
        builder.addContentView(imageView, new RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
        builder.show();
    }
}

File: TiAires/app/src/main/java/net/teamtruta/tiaires/views/GeoCacheDetailActivity.kt
package net.teamtruta.tiaires.views

import android.app.Dialog
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.*
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.FileProvider
import com.microsoft.appcenter.analytics.Analytics
import net.teamtruta.tiaires.App
import net.teamtruta.tiaires.R
import net.teamtruta.tiaires.data.models.GeoCacheInTour
import net.teamtruta.tiaires.data.models.GeoCacheInTourWithDetails
import net.teamtruta.tiaires.data.models.VisitOutcomeEnum
import net.teamtruta.tiaires.viewModels.GeoCacheDetailViewModel
import net.teamtruta.tiaires.viewModels.GeoCacheDetailViewModelFactory
import org.honorato.multistatetogglebutton.MultiStateToggleButton
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class GeoCacheDetailActivity : AppCompatActivity(), PopupMenu.OnMenuItemClickListener {

    private var currentGeoCacheInTour: GeoCacheInTour? = null
    private var soundPool: SoundPool? = null
    private var soundID = 0
    private var currentPhotoPath: String = ""
    private var TAG: String = GeoCacheDetailActivity::class.java.simpleName

    private val viewModel: GeoCacheDetailViewModel by viewModels{
        GeoCacheDetailViewModelFactory((application as App).repository)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_geo_cache_detail)
        val toolbar = findViewById<Toolbar>(R.id.toolbar_geo_cache_detail)
        setSupportActionBar(toolbar)
        val geoCacheID = intent.getLongExtra(App.GEOCACHE_ID_EXTRA, -1L)

        // Setup Action Bar
        val ab = supportActionBar!!
        ab.setDisplayHomeAsUpEnabled(true)

        // Observe geocache that was clicked on
        viewModel.getGeoCacheInTourFromID(geoCacheID).observe(this,
                { geoCacheInTour: GeoCacheInTourWithDetails -> setupCacheSpecificData(geoCacheInTour) })


        //  Setup ping sound
        setupAudio()
    }

    private fun setupCacheSpecificData(geoCacheInTour: GeoCacheInTourWithDetails) {
        currentGeoCacheInTour = geoCacheInTour.geoCacheInTour
        val currentGeoCache = geoCacheInTour.geoCache.geoCache

        // Set GeoCache Title
        val ab = supportActionBar
        ab!!.title = currentGeoCache.name


        // Set Not Found / Found / DNF toggle and appropriate onClickListener
        val geoCacheVisitButton = findViewById<MultiStateToggleButton>(R.id.geo_cache_visit_button)
        var buttonStates = booleanArrayOf(true, false, false)
        if (currentGeoCacheInTour!!.currentVisitOutcome === VisitOutcomeEnum.Found) buttonStates = booleanArrayOf(false, true, false) else if (currentGeoCacheInTour!!.currentVisitOutcome === VisitOutcomeEnum.DNF) buttonStates = booleanArrayOf(false, false, true)
        geoCacheVisitButton.states = buttonStates

        geoCacheVisitButton.setOnValueChangedListener { position: Int ->
            when(position){
                0 -> viewModel.setGeoCacheInTourVisit(currentGeoCacheInTour!!, VisitOutcomeEnum.NotAttempted)
                1 -> {
                    viewModel.setGeoCacheInTourVisit(currentGeoCacheInTour!!, VisitOutcomeEnum.Found)
                    playPing()}
                2-> {viewModel.setGeoCacheInTourVisit(currentGeoCacheInTour!!, VisitOutcomeEnum.DNF)
                    playPing()}
            }
        }

        // Set Checkboxes
        // 1. Needs Maintenance Checkbox
        val needsMaintenanceCheckBox = findViewById<CheckBox>(R.id.needsMaintenanceCheckBox)
        needsMaintenanceCheckBox.isChecked = currentGeoCacheInTour!!.needsMaintenance

        // 2. FoundTrackable Checkbox and EditText
        val foundTrackableCheckBox = findViewById<CheckBox>(R.id.foundTrackableCheckBox)
        if (currentGeoCacheInTour!!.foundTrackable != null) {
            foundTrackableCheckBox.isChecked = true
            val foundTrackableEditText = findViewById<EditText>(R.id.foundTrackableEditText)
            foundTrackableEditText.setText(currentGeoCacheInTour!!.foundTrackable)
        }

        // 3. DroppedTrackable Checkbox and EditText
        val droppedTrackableCheckBox = findViewById<CheckBox>(R.id.droppedTrackableCheckBox)
        if (currentGeoCacheInTour!!.droppedTrackable != null) {
            droppedTrackableCheckBox.isChecked = true
            val droppedTrackableEditText = findViewById<EditText>(R.id.droppedTrackableEditText)
            droppedTrackableEditText.setText(currentGeoCacheInTour!!.droppedTrackable)
        }

        // 4. Favourite Point Checkbox
        val favouritePointCheckBox = findViewById<CheckBox>(R.id.favouritePointCheckBox)
        if (currentGeoCacheInTour!!.favouritePoint) favouritePointCheckBox.isChecked = true

        // Set my notes
        val notesSection = findViewById<EditText>(R.id.notes)
        val myNotes = currentGeoCacheInTour!!.notes
        if (myNotes != "") notesSection.setText(myNotes)

        // Setup photo
        val photoCheckBox = findViewById<CheckBox>(R.id.photo_checkbox)
        photoCheckBox.isChecked = currentGeoCacheInTour!!.pathToImage != null
    }

    override fun onSupportNavigateUp(): Boolean {
        val intent = Intent(this, TourActivity::class.java)

        setResult(RESULT_OK, intent)
        finish()
        return true
    }

    override fun onBackPressed() {
        saveChanges()
        onSupportNavigateUp()
    }


    /**
     * Handle Saving in the geoCache in tour activity
     */
    private fun saveChanges() {
        Analytics.trackEvent("GeoCacheDetailActivity.saveChanges")

        // save changes
        val needsMaintenanceCheckBox = findViewById<CheckBox>(R.id.needsMaintenanceCheckBox)
        val favouritePointCheckBox = findViewById<CheckBox>(R.id.favouritePointCheckBox)
        val foundTrackableEditText = findViewById<EditText>(R.id.foundTrackableEditText)
        val droppedTrackableEditText = findViewById<EditText>(R.id.droppedTrackableEditText)
        val notesView = findViewById<EditText>(R.id.notes)

        viewModel.updateGeocaCheInTourDetails(currentGeoCacheInTour!!,
                needsMaintenanceCheckBox.isChecked,
                favouritePointCheckBox.isChecked,
                foundTrackableEditText.text.toString(),
                droppedTrackableEditText.text.toString(),
                notesView.text.toString())
    }


    fun onFoundTrackableCheckboxClicked(view: View?) {
        val foundTrackableCheckBox = findViewById<CheckBox>(R.id.foundTrackableCheckBox)
        val foundTrackableEditText = findViewById<EditText>(R.id.foundTrackableEditText)
        if (foundTrackableCheckBox.isChecked) {
            // Focus attention on text box to fill in value
            foundTrackableEditText.requestFocus()
        } else {
            // Delete inputted trackable code in editText area
            foundTrackableEditText.setText("")
        }
    }

    fun onDroppedTrackableCheckboxClicked(view: View?) {
        val droppedTrackableCheckBox = findViewById<CheckBox>(R.id.droppedTrackableCheckBox)
        val droppedTrackableEditText = findViewById<EditText>(R.id.droppedTrackableEditText)
        if (droppedTrackableCheckBox.isChecked) {
            // Focus attention on text box to fill in value
            droppedTrackableEditText.requestFocus()
        } else {
            droppedTrackableEditText.setText("")
        }
    }


    override fun onPause() {
        saveChanges()
        super.onPause()
    }

    fun playPing() {
        // AudioManager audio settings for adjusting the volume
        val audioManager = getSystemService(AUDIO_SERVICE) as AudioManager
        val volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC).toFloat()
        soundPool!!.play(soundID, volume, volume, 1, 0, 1f)
    }

    private fun setupAudio() {
        val attributes = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_GAME)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build()
        soundPool = SoundPool.Builder()
                .setAudioAttributes(attributes)
                .build()
        soundID = soundPool?.load(this, R.raw.ping, 1)!!
    }

    fun takePhoto(view: View?) {
        takePhoto()
    }

    private fun takePhoto() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(packageManager) != null) {
            // Create the File where the photo should go
            var photoFile: File? = null
            try {
                photoFile = createImageFile()
            } catch (ex: IOException) {
                // TODO: Error occurred while creating the File
                ex.printStackTrace()
            }
            // Continue only if the File was successfully created
            if (photoFile != null) {
                val photoURI = FileProvider.getUriForFile(this,
                        "net.teamtruta.tiaires.fileprovider",
                        photoFile)
                Log.d(TAG, photoURI.toString())
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == RESULT_OK) {
            galleryAddPic()
        }
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        // Create an image file name
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
                .format(Date())
        val imageFileName = "GEOCACHEID_" + currentGeoCacheInTour!!.id + "_" + timeStamp + "_"
        val storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        val image = File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",  /* suffix */
                storageDir /* directory */
        )

        // Save a file: path for use with ACTION_VIEW intents
        currentPhotoPath = image.absolutePath
        currentGeoCacheInTour!!.pathToImage = currentPhotoPath
        return image
    }

    private fun galleryAddPic() {
        //invoke the system's media scanner to add your photo to the Media Provider's database,
        // making it available in the Android Gallery application and to other apps.
        val mediaScanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
        val f = File(currentPhotoPath)
        val contentUri = Uri.fromFile(f)
        mediaScanIntent.data = contentUri
        this.sendBroadcast(mediaScanIntent)
    }

    fun showPhotoPopup(view: View?) {

        // If the current geoCache has no photo then go straight into the photo taking
        if (currentGeoCacheInTour!!.pathToImage == null) {
            takePhoto()
            // Set the checkbox to checked
            val photoCheckBox = findViewById<CheckBox>(R.id.photo_checkbox)
            photoCheckBox.isChecked = true
            return
        }

        // Set the checkbox to checked
        val photoCheckBox = findViewById<CheckBox>(R.id.photo_checkbox)
        photoCheckBox.isChecked = true
        val popup = PopupMenu(this, view)
        // This activity implements OnMenuItemClickListener
        popup.setOnMenuItemClickListener(this)
        popup.inflate(R.menu.photo_actions)
        popup.show()
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.take_new_photo) {
            takePhoto()
            return true
        }
        if (id == R.id.show_geo_cache_photo) {
            showGeoCachePhoto()
            return true
        }
        if (id == R.id.delete_geo_cache_photo) {
            deleteGeoCachePhoto()
            return true
        }
        return false
    }

    private fun deleteGeoCachePhoto() {
        val file = File(Objects.requireNonNull(currentGeoCacheInTour!!.pathToImage))
        if (file.exists()) {
            if (file.delete()) {
                println("File Deleted :")
            } else {
                println("File not Deleted :")
            }
        }
        currentGeoCacheInTour!!.pathToImage = null
        val photoCheckbox = findViewById<CheckBox>(R.id.photo_checkbox)
        photoCheckbox.isChecked = false
    }

    /*  // Throws an exception because it exposes App files to the system
    void showCachePhoto(){
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_VIEW);
        intent.setDataAndType(Uri.parse("file://" + currentGeoCache.getPathToImage()), "image/ *");
        startActivity(intent);
    }
*/
    private fun showGeoCachePhoto() {
        val builder = Dialog(this)
        builder.requestWindowFeature(Window.FEATURE_NO_TITLE)
        builder.window!!.setBackgroundDrawable(
                ColorDrawable(Color.TRANSPARENT))
        builder.setOnDismissListener { }
        val imageView = ImageView(this)
        imageView.setImageURI(Uri.parse(currentGeoCacheInTour!!.pathToImage))
        builder.addContentView(imageView, RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT))
        builder.show()
    }

    companion object {
        const val REQUEST_IMAGE_CAPTURE = 1
    }
}